pkg/statemetrics/xr_state_metrics.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+)
+
+// A XRStateRecorder records the state of composite resources.
+type XRStateRecorder struct {
+	client    client.Client
+	log       logging.Logger
+	frequency time.Duration
+
+	compositeExists        *prometheus.GaugeVec
+	compositeReady         *prometheus.GaugeVec
+	compositeSynced        *prometheus.GaugeVec
+	compositeComposedCount *prometheus.GaugeVec
+}
+
+// A APIExtStateRecorderOption configures a MRStateRecorder.
+type APIExtStateRecorderOption func(*XRStateRecorder)
+
+// NewXRStateRecorder returns a new XRStateRecorder which records the state of claim,
+// composite and composition metrics.
+func NewXRStateRecorder(client client.Client, log logging.Logger, frequency time.Duration) *XRStateRecorder {
+	return &XRStateRecorder{
+		client:    client,
+		log:       log,
+		frequency: frequency,
+
+		compositeExists: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      "composite_resource_exists",
+			Help:      "The number of composite resources that exist",
+		}, []string{"gvk", "composition"}),
+		compositeReady: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      "composite_resource_ready",
+			Help:      "The number of composite resources in Ready=True state",
+		}, []string{"gvk", "composition"}),
+		compositeSynced: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      "composite_resource_synced",
+			Help:      "The number of composite resources in Synced=True state",
+		}, []string{"gvk", "composition"}),
+		compositeComposedCount: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      "composite_resource_composed_count",
+			Help:      "The number of composed resources in the XR",
+		}, []string{"gvk", "composition"}),
+	}
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+// collected by this Collector to the provided channel and returns once
+// the last descriptor has been sent.
+func (r *XRStateRecorder) Describe(ch chan<- *prometheus.Desc) {
+	r.compositeExists.Describe(ch)
+	r.compositeReady.Describe(ch)
+	r.compositeSynced.Describe(ch)
+	r.compositeComposedCount.Describe(ch)
+}
+
+// Collect is called by the Prometheus registry when collecting
+// metrics. The implementation sends each collected metric via the
+// provided channel and returns once the last metric has been sent.
+func (r *XRStateRecorder) Collect(ch chan<- prometheus.Metric) {
+	r.compositeExists.Collect(ch)
+	r.compositeReady.Collect(ch)
+	r.compositeSynced.Collect(ch)
+	r.compositeComposedCount.Collect(ch)
+}
+
+// Record records the state of managed resources.
+func (r *XRStateRecorder) Record(ctx context.Context, gvk schema.GroupVersionKind) {
+	l := &unstructured.UnstructuredList{}
+	l.SetGroupVersionKind(gvk)
+	err := r.client.List(ctx, l)
```

Comment:
Don't list from a client. List from a cache, i.e. pass in a cache here. Listing can be expensive, imagine 10000 objects.